packages/tpetra/core/src/Tpetra_Details_DistributorActor.hpp

```diff
@@ -0,0 +1,796 @@
+        //    packets from process getProcsFrom()[i]).
+        // 2. Start the Irecv and save the resulting request.
+        TEUCHOS_TEST_FOR_EXCEPTION(
+            curBufferOffset + curBufLen > static_cast<size_t> (imports.size ()),
+            std::logic_error, "Tpetra::Distributor::doPosts(3 args, Kokkos): "
+            "Exceeded size of 'imports' array in packing loop on Process " <<
+            myRank << ".  imports.size() = " << imports.size () << " < "
+            "curBufferOffset(" << curBufferOffset << ") + curBufLen(" <<
+            curBufLen << ").");
+        imports_view_type recvBuf =
+          subview_offset (imports, curBufferOffset, curBufLen);
+        requests_.push_back (ireceive<int> (recvBuf, plan.getProcsFrom()[i],
+              tag, *plan.getComm()));
+      }
+      else { // Receiving from myself
+        selfReceiveOffset = curBufferOffset; // Remember the self-recv offset
+      }
+      curBufferOffset += curBufLen;
+    }
+  }
+
+  if (doBarrier) {
+#ifdef HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+    Teuchos::TimeMonitor timeMonBarrier (*timer_doPosts3KV_barrier_);
+#endif // HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+
+    // If we are using ready sends (MPI_Rsend) below, we need to do
+    // a barrier before we post the ready sends.  This is because a
+    // ready send requires that its matching receive has already
+    // been posted before the send has been posted.  The only way to
+    // guarantee that in this case is to use a barrier.
+    plan.getComm()->barrier ();
+  }
+
+#ifdef HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+  Teuchos::TimeMonitor timeMonSends (*timer_doPosts3KV_sends_);
+#endif // HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+
+  // setup scan through getProcsTo() list starting with higher numbered procs
+  // (should help balance message traffic)
+  //
+  // FIXME (mfh 20 Feb 2013) Why haven't we precomputed this?
+  // It doesn't depend on the input at all.
+  size_t numBlocks = plan.getNumSends() + plan.hasSelfMessage();
+  size_t procIndex = 0;
+  while ((procIndex < numBlocks) && (plan.getProcsTo()[procIndex] < myRank)) {
+    ++procIndex;
+  }
+  if (procIndex == numBlocks) {
+    procIndex = 0;
+  }
+
+  size_t selfNum = 0;
+  size_t selfIndex = 0;
+
+  if (plan.getIndicesTo().is_null()) {
+
+#ifdef HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+    Teuchos::TimeMonitor timeMonSends2 (*timer_doPosts3KV_sends_fast_);
+#endif // HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+
+    // Data are already blocked (laid out) by process, so we don't
+    // need a separate send buffer (besides the exports array).
+    for (size_t i = 0; i < numBlocks; ++i) {
+      size_t p = i + procIndex;
+      if (p > (numBlocks - 1)) {
+        p -= numBlocks;
+      }
+
+      if (plan.getProcsTo()[p] != myRank) {
+        exports_view_type tmpSend = subview_offset(
+            exports, plan.getStartsTo()[p]*numPackets, plan.getLengthsTo()[p]*numPackets);
+
+        if (sendType == Details::DISTRIBUTOR_SEND) {
```

Comment:
Eventually (not this PR), we should decide whether to remove all these different send types.  Are they needed?

Reply:
agreed.  added to my list for future work